engine/gamesys/src/gamesys/test/test_gamesys.cpp

```diff
@@ -447,17 +447,16 @@ TEST_F(SoundTest, UpdateSoundResource)
     ASSERT_TRUE(dmGameObject::Update(m_Collection, &m_UpdateContext));
     ASSERT_TRUE(dmGameObject::PostUpdate(m_Collection));
 
-    // Check the size of the updated resource. Reports +16 for some reason
+    // Check the size of the updated resource
 
     dmhash_t comp_name = dmHashString64("dynamic-sound"); // id of soundc component
     dmhash_t prop_name = dmHashString64("sound"); // property of sound data resource within a sound component
     // Get hash of the sounddata resource
     dmhash_t soundata_hash = 0;
     GetResourceProperty(go, comp_name, prop_name, &soundata_hash);
-    //ASSERT_EQ(dmHashString64("/sound/valid.wav"), soundata_hash);  // for some reason the following asssertion fails. What should be the actual soundata path (before hashing) ?
     dmResource::SResourceDescriptor* descp = dmResource::FindByHash(m_Factory, soundata_hash);
     dmLogInfo("size: %i", descp->m_ResourceSize);
-    ASSERT_EQ(98510+16, descp->m_ResourceSize);  // TOCHECK: size returned is always +16 from size of wav. Why ?
+    ASSERT_EQ(98510+16, descp->m_ResourceSize);  // size returned is always +16 from size of wav: sound_data->m_Size + sizeof(SoundData);
```

Comment:
I recommend mosing the `struct SoundData` to the `sound_private.h`. That way you can use `sizeof(SoundData)` to make the test a bit more robust.

Reply:
@JCash 'mosing' wdym ? :-)

`struct SoundData` is _defined_ in sound.cpp. As far as i understand, test_gamesys.cpp requires the definition to be present in order to determine the sizeof it. Not the mere declaration.

We could move the whole SoundData definition somewhere else where it could be available in test_gamesys.cpp? Maybe sound_private.h? And include sound_private.h in test_gamesys.cpp. Is that what you imply ?

hmmm,..If that's the case, i think i start to understand the use of the _private suffix in header files. I.e. do not expose a type definition but make it available in safe places where it's really needed like tests.

Reply:
> We could move the whole SoundData definition somewhere else where it could be available in test_gamesys.cpp? Maybe sound_private.h? And include sound_private.h in test_gamesys.cpp. Is that what you imply ?

>hmmm..If that's the case, i think i start to understand the use of the _private suffix in header files. I.e. do not expose a type definition but make it available in safe places where it's really needed like tests.

Precisely!

Reply:
@JCash , good and bad news.

I followed the plan we discussed above. However i ran into some trouble.

In test_gamesys.cpp, just printing the sizeof(SoundData) returned 32 (8-byte granular size of SoundData moved to sound_private.h). However, `dmResource::SResourceDescriptor* descp -> m_ResourceSize) kept returning "actual size + 16".

So, it seems that moving SoundData to sound_private.h was wrong. This structure is not really used in the test. Instead, there is another `SoundData` structure defined in sound_null.cpp which fits the bill. It's size is  indeed 16.

What i finally did was to externalize the SoundData structure defined in sound_null.cpp to a new file named sound_null_private.h. 

Tell me if you're ok with this approach.

Reply:
From [sound.cpp](https://github.com/defold/defold/blob/dev/engine/sound/src/sound.cpp#L501-L504)
```
    uint32_t GetSoundResourceSize(HSoundData sound_data)
    {
        return sound_data->m_Size + sizeof(SoundData);
    }
```

This is the actual calculation, and it uses the `struct SoundData` from line [146-154](https://github.com/defold/defold/blob/dev/engine/sound/src/sound.cpp#L146-L154).

The `sound.cpp` is used in those tests that have sounds actually playing, `sound_null.cpp` in those that will play thorugh  a null device.

> What i finally did was to externalize the SoundData structure defined in sound_null.cpp to a new file named sound_null_private.h.

I'd say it's even more dangerous to make only one of the two structs available in the private header given its different sizes in different files.

Ok, so what I propose instead, is that we go back to what you had previously, with a hardcoded constant 16, but with a comment that the 16 refers to the sizeof(SoundData) from sound_null.cpp. That way it will be easy to fix if the struct size will change (and the test starts failing).

Reply:
cool. Let me revert to the previous then.